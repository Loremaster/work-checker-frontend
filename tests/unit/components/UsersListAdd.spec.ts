import { mount, createLocalVue } from "@vue/test-utils";
import { insertDataAppDiv } from "../helpers/vuetify-helper";
import Vuetify from "vuetify";
import Vue from "vue";
import Vuex from "vuex";
import flushPromises from "flush-promises";
import Component from "@/components/UsersListAdd.vue";
import VeeValidate from "vee-validate";

const localVue = createLocalVue();
const $validator = new VeeValidate.Validator();
const email = "user@email.com";
const password = "password12345";
const role = "admin";

localVue.use(Vuex);

Vue.use(Vuetify);
Vue.use(VeeValidate);

let actions: any;
let store: any;
let wrapper: any;

describe("UsersListAdd.vue", () => {
  beforeEach(() => {
    actions = { createUser: jest.fn() };
  });

  const setup = () => {
    insertDataAppDiv();
    store = new Vuex.Store({ actions });

    wrapper = mount(Component, {
      store,
      localVue,
      sync: false,
      provide: () => ({ $validator }),
    });
  };

  it("renders correctly", () => {
    setup();

    expect(wrapper).toMatchSnapshot();
  });

  it("submits form", async () => {
    setup();

    await submit({ email, password, role });

    expect(actions.createUser).toHaveBeenCalledWith(expect.anything(), { email, password, role }, undefined);
  });

  it("emits created event", async () => {
    setup();

    await submit({ email, password, role });

    expect(wrapper.emitted().created).toBeTruthy();
  });

  it("renders server errors", async () => {
    const error: any = new Error("Async error");
    error.response = { data: { email: ["must be unique"] } };
    actions.createUser = jest.fn().mockRejectedValue(error);
    setup();

    await submit({ email, password, role });

    expect(wrapper.html()).toContain("must be unique");
  });
});

async function submit({ email, password, role }: { email: string; password: string; role: string }) {
  wrapper.find("input[name='email']").setValue(email);
  wrapper.find("input[name='password']").setValue(password);
  wrapper.find("input[name='role']").setValue(role);
  wrapper.find({ ref: "createBtn" }).trigger("click");

  await flushPromises();
}
