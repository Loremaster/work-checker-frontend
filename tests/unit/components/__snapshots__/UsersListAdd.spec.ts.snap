// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersListAdd.vue renders correctly 1`] = `
<form novalidate="novalidate" class="v-form">
  <div class="container">
    <div class="layout">
      <div class="flex md4">
        <div class="v-input v-text-field theme--light">
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot"><label aria-hidden="true" class="v-label theme--light" style="left: 0px; position: absolute;">Email</label><input aria-label="Email" name="email" data-vv-name="email" required="required" type="text"></div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <div class="v-messages__wrapper"></div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="flex md3">
        <div class="v-input v-text-field theme--light">
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot"><label aria-hidden="true" class="v-label theme--light" style="left: 0px; position: absolute;">Password</label><input aria-label="Password" name="password" data-vv-name="password" required="required" type="text"></div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <div class="v-messages__wrapper"></div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="flex md3">
        <div role="combobox" class="v-input v-text-field v-select theme--light">
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-select__slot"><label aria-hidden="true" class="v-label theme--light" style="left: 0px; position: absolute;">Role</label>
                <div class="v-select__selections"><input aria-label="Role" name="role" data-vv-name="role" required="required" readonly="readonly" type="text" autocomplete="on" aria-readonly="false"></div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon material-icons theme--light">arrow_drop_down</i></div>
                </div>
              </div>
              <div class="v-menu"></div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--light">
                <div class="v-messages__wrapper"></div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="flex md1"><button type="button" class="v-btn theme--light">
          <div class="v-btn__content">Cancel</div>
        </button></div>
      <div class="flex md1"><button type="button" class="v-btn theme--light primary">
          <div class="v-btn__content">Create</div>
        </button></div>
    </div>
  </div>
</form>
`;
